### PR TITLE
refactor: email test cases

### PR DESCRIPTION
**Pull request by ariskemper**
@fabian-hiller Here is my proposal for email test cases. I fixed regex so it supports
max 64 chars in local part of email value and 255 chars in domain part.

### CODE REVIEW

**Comment by ariskemper** [comment]
@fabian-hiller We could wrap the tests in another describe for grouping. I was also thinking to create a helper function which takes as an argument validate._parse(value).issues or validate._parse(value).output and returns true or false.
Would be nicer to read test like 

`expect(isValidParse(validate._parse(value))).toBeTruthy();`

or

`expect(isValidParse(validate._parse(value))).toBeFalsy();`

**Comment by fabian-hiller** [comment]
I think I like the grouping but I am not sure about the utility functions. I like that the tests show the raw code, but on the other hand I also see the advantage of making the code more readable and simple. If we were to create a helper function for this, I would choose the following API:

```ts
expect(isValidParse(validate, value)).toBeTruthy();
expect(isValidParse(validate, value)).toBeFalsy();
```

Do you have an idea for alternative names for the utility function? Maybe `passes`, `validate` or `parse`. I am looking for a word that makes the whole API more readable.

**Comment by ariskemper** [comment]
@fabian-hiller i also like tests showing raw code, but readable tests are also important so my suggestion is following:

```
const emailValidate = email();
expect(parse(emailValidate, value)).toBeTruthy();
expect(parse(emailValidate, value)).toBeFalsy();
```

**Comment by fabian-hiller** [comment]
Looks good to me. Feel free to implement this for some tests so we can compare. A disadvantage of this approach is that not everything can be covered with `true` and `false` and we still have to resort to raw code in some cases.

File: library/src/regex.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ export const CUID2_REGEX = /^[a-z][\da-z]*$/u;
  * Email regex.
  */
 export const EMAIL_REGEX =
-  /^[\w+-]+(?:\.[\w+-]+)*@[\da-z]+(?:[.-][\da-z]+)*\.[a-z]{2,}$/iu;
+  /^[\w+-]{1,64}(?:\.[\w+-]{1,64})*@[\da-z]{1,255}(?:[.-][\da-z]{1,255})*\.[a-z]{2,}$/iu;
```
[the remark this answers is not in the document]
Comment (fabian-hiller) [reply library/src/regex.ts:10]: Our regex deliberately tries not to conform to the standard. However, there is a discussion in #204 about whether we want to add a separate validation function for this.

[the remark this answers is not in the document]
Comment (ariskemper) [reply library/src/regex.ts]: @fabian-hiller i reverted email Regex and removed associated tests. In the future, it could be changed, if agreed.

**Comment by ariskemper** [comment]
@fabian-hiller what do you think about this PR and parseValidation? If you agree, we could merge it and i refactor current mac and bic tests?

**Comment by ariskemper** [comment]
@fabian-hiller any update on this?

**Comment by fabian-hiller** [comment]
It's on my list. I have not forgotten. By the end of next week I will be working full time on Valibot for the next few weeks. I expect big progress on the source code and documentation.

**Comment by ariskemper** [comment]
@fabian-hiller i was thinking maybe that we could have some status in validation, which is setled, based on validation true, false, besides issues array.

**Comment by fabian-hiller** [comment]
What is the advantage of such an status key in your eyes? So far I have consciously decided against it for the internal code because it was not necessary. For external use of a schema, for example with `safeParse`, such a status is returned with `.success`.

**Review by fabian-hiller — CHANGES_REQUESTED** [review]
I like the more fine-grained tests with a separate message for each case. I also like the 3-layer system (the grouping) with the nested `describe`. I am not sure about the `parseValidation` function, because I think it is better not to add an abstraction here. Currently I tend not to use it. Also, instead of just testing `.output`, I think that I would prefer if we test the entire object returned by `email`. Feel free to give me feedback on this.

**Comment by fabian-hiller** [comment]
Thank you again for your contribution. I will try to review it the next days and give you feedback.

**Review by fabian-hiller — CHANGES_REQUESTED** [review]
This PR looks good now. There is only one thing I would change. I think instead of just testing the output, we should test the whole object returned by `._parse`. This makes the tests safer. What do you think?

```ts
// Old
test('should pass standard email', () => {
  const value = 'email@example.com';
  expect(validate._parse(value).output).toBe(value);
});

// New
test('should pass standard email', () => {
  const value = 'email@example.com';
  expect(validate._parse(value)).toEqual({ output: value });
});
```

**Comment by fabian-hiller** [comment]
Once this PR is merged, I recommend creating another PR for the tests of the `string` schema function, and then we can rework the tests of the entire library.

**Comment by ariskemper** [comment]
@fabian-hiller i agree to the recommendation to compare with output object.

**Comment by fabian-hiller** [comment]
Thanks for updating this PR! I will review it again next week and probably merge it.

**Comment by ariskemper** [comment]
@fabian-hiller please check PR https://github.com/fabian-hiller/valibot/pull/366 for string schema refactored tests.

**Comment by fabian-hiller** [comment]
Please have a look an review my changes. Note: I have removed the "should return custom error message" test because after the refactor of #397 we will pass the [entire context](https://github.com/fabian-hiller/valibot/blob/feat-i18n/library/src/validations/email/email.ts#L39) to `actionIssue`. With "should contain the message argument" we ensure that the messages is part of the context.

**Review by fabian-hiller — APPROVED** [review]

File: library/src/validations/email/email.test.ts [hunk 2]
```diff
@@ -2,60 +2,200 @@ import { describe, expect, test } from 'vitest';
...
-    ).toBeTruthy();
-    expect(validate._parse('email@example').issues).toBeTruthy();
-    expect(validate._parse('email@-example.com').issues).toBeTruthy();
-    expect(validate._parse('email@example-.com').issues).toBeTruthy();
-    expect(validate._parse('email@111.222.333.44444').issues).toBeTruthy();
-    expect(validate._parse('email@example..com').issues).toBeTruthy();
-    expect(validate._parse('Abc..123@example.com').issues).toBeTruthy();
+  describe('should return action object', () => {
+    test('should contain the default message', () => {
+      expect(email()).toMatchObject({
```
Comment (fabian-hiller) [inline library/src/validations/email/email.test.ts:7]: Is `.toMatchObject()` less strict than `.toEqual()`? Why do you choose it here?

**Comment by fabian-hiller** [comment]
Thank you again for this PR Aris! I added your tests to the rewrite of Valibot in this [commit](https://github.com/fabian-hiller/valibot/pull/502/commits/8023fea347c266f604285f61736b40729195d4cc).

**Pull request closed without merging**